python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py

```diff
@@ -978,66 +961,38 @@ def fetch_row_counts(
                 A set of corresponding Dagster events for dbt models, with row counts attached,
                 yielded in the order they are emitted by dbt.
         """
-        return DbtEventIterator(
-            self._fetch_row_counts_inner(),
-            dbt_cli_invocation=self._dbt_cli_invocation,
-        )
+        # If the adapter is DuckDB, we need to wait for the dbt CLI process to complete
+        # so that the DuckDB lock is released. This is because DuckDB does not allow for
+        # opening multiple connections to the same database when a write connection, such
+        # as the one dbt uses, is open.
+        block_on_dbt_run = False
+        try:
+            from dbt.adapters.duckdb import DuckDBAdapter
 
-    def _fetch_row_counts_inner(
-        self,
-    ) -> Iterator[
-        Union[
-            Output,
-            AssetMaterialization,
-            AssetObservation,
-            AssetCheckResult,
-        ]
-    ]:
-        logger.info("Row counts will be fetched for non-view models once they are materialized.")
-
-        # We keep a list of emitted Dagster events and pending futures which augment
-        # emitted events with additional metadata. This ensures we can yield events in the order
-        # they are emitted by dbt.
-        output_events_and_futures: List[Union[Future, DbtDagsterEventType]] = []
-
-        # Point at project directory to ensure dbt adapters run correctly
-        with pushd(str(self._dbt_cli_invocation.project_dir)), ThreadPoolExecutor(
-            max_workers=STREAM_EVENTS_THREADPOOL_SIZE
-        ) as executor:
-            self._dbt_run_thread = executor.submit(
-                self._stream_dbt_events_and_enqueue_postprocessing,
-                output_events_and_futures,
-                executor,
-            )
+            if isinstance(self._dbt_cli_invocation.adapter, DuckDBAdapter):
+                block_on_dbt_run = True
+        except ImportError:
+            pass
 
-            # Step through the list of output events and futures, yielding them in order
-            # once they are ready to be emitted
-            event_to_emit_idx = 0
-            while True:
-                all_work_complete = get_future_completion_state_or_err(
-                    [self._dbt_run_thread, *output_events_and_futures]
+        def _threadpool_fetch_and_attach_row_count_metadata() -> (
+            Iterator[Union[Output, AssetMaterialization, AssetObservation, AssetCheckResult]]
+        ):
+            with ThreadPoolExecutor(
+                max_workers=self._dbt_cli_invocation.postprocessing_threadpool_num_threads,
+                thread_name_prefix="fetch_row_counts_",
```

Comment:
I believe there's already a separator added, so no need to add that manually

```suggestion
                thread_name_prefix="fetch_row_counts",
```